### PR TITLE
[Feature]: "Core" Update for Dotly

### PR DESCRIPTION
**Pull request by gtrabanco**
# Tasks
* [x] Check no BASH syntax problems
* [x] Make some new functions tests
* [x] Added async library
* [x] Testing done, ready to merge

# Testing tasks
* [x] Testing autoloader
* [x] Testing async autoupdater
* [x] Check new installation
* [x] Check if vars were replacing
* [x] Check if autoupdater can be enabled/disabled/check status with:
> * [x] `dot self init enable`
> * [x] `dot self init enable dotly_autoupdater`
> * [x] `dot self init disable`
> * [x] `dot self init disable dotly_autoupdater`
> * [x] `dot self init status`
> * [x] `dot self init status dotly_autoupdater`

# Features

* Added again PR #61 because it is necessary when I use `beta` branch for testing.
* Now the dotly core loader is outside of `.bashrc` and `.zshrc` files so now the framework core functionality could be modified.
* Added optional initializer scripts after the DOTLY framework is loaded, to load, for example the autoupdater. You need to enable it manually by executing `dot self init enable`. To disable init scripts just type: `dot self init disable` you can see the status of `autoupdate_loader` with `dot self init status autoupdate_loader` or `dot self init status`. See more usage with `dot self init -h`. #112 
* Added some functions that are used in other PRs. Templating (new), debug (new), array (new) str, output and platform.
* Added optional library (should be included manually) to do some debug when building scripts. Not sure if this is for a DOTLY.
* `PATH`s are now outside of `exports.sh` and it is loaded at the very beginning.

# Motivation

This is done because there are some recurrent functions that were used in my other PRs. So merge with master later could be a mess, with this I try to simplify the process of merging the other PRs because there are a lot stuff.

# Important

This PR will be included in other PRs like dotly_marketplace (PR #116). It is important update that will change how dotly run at startup, it is not a very big change but it is important and needs testing before apply.

### CODE REVIEW

**Comment by gtrabanco** [comment]
* Added another task in main message
* Added `#Important` title in main message.

**Comment by gtrabanco** [comment]
* Added str::to_upper
* Added platform functions to get current script full path or directory (useful to include libs)

**Comment by gtrabanco** [comment]
Tested and ready to merge, anyway one fast test in other computer may help.
* Async library were added to make some improvements to branch `feature/marketplace_extended`
* Autoupdate run synchronously, I will re-make it async when I have finished `feature/marketplace_extended` and `feature/ssh_config`.

**Comment by gtrabanco** [comment]
* `dot symlinks apply update` subcommand that was added was splitted. This was done becasue the use case is different for this subcommand than `dot symlinks apply`. It will also make easier apply later PR #79 that "solves" #76.

@rgomezcasas @JavierCane this PR is ready to merge.

**Comment by gtrabanco** [comment]
* Added async in `_main.sh` core loading

Would be nice having a reference in README to the creator repository of the async library in bash:
* https://github.com/zombieleet/async-bash

**Comment by gtrabanco** [comment]
* Fixed error 93, I could not tested on bash without applying this patch that is also in PR #95 and reported in #93
* Fixed Autoload on init scripts
* Now autoupdate is async (was easier begin with async here).

# Known errors
* The async function of async library, print 1 on screen every time a async function is called.
* The async functions previous to be called in scripts need to do `set +eu` or it will fail.
* Problem? With prompt. See Help needed.

# How to use autoupdate

To enable autoupdate (not by default), type:
```bash
dot self init enable dotly_autoupdater
```

There are three modes of dotly autoupdater which is defined in the value of ENV var `DOTLY_AUTO_UPDATE_MODE`:
* `DOTLY_AUTO_UPDATE_MODE=auto`: It will autoupdate dotly
* `DOTLY_AUTO_UPDATE_MODE=info`: default mode.
> * It will show an advise in your terminal and create a file with the path: `$DOTFILES_PATH/.dotly_update_available`
> * This file is used to show you a box in the PROMPT.
* `DOTLY_AUTO_UPDATE_MODE=prompt`: no advise showed, just create the file to see the box in the prompt.

You can also define the number of days that your `DOTLY_PATH` should be older to begin check for updates:

```bash
DOTLY_AUTO_UPDATE_DAYS=15
```

If you do not define anyone these are the default values:

```bash
DOTLY_AUTO_UPDATE_DAYS=7
DOTLY_AUTO_UPDATE_MODE=info
```

As I said, this won't work unless you enable it first:
```bash
dot self init enable dotly_autoupdater
```

To disable:
```bash
dot self init disable dotly_autoupdater
```

# Help needed

* I need help with the prompt to make the box disappear if I delete the file without reseting the terminal.
* Also help needed to make the box appear. Do not know why the prompt is not updating after press the key "intro". Should not the prompt update itself?

**Comment by gtrabanco** [comment]
Sorry for the last two commits but I merged `feature/ssh_config` into this. Now should be solved.

**Comment by gtrabanco** [comment]
Ready knowing the known issues and help needed with prompt.

**Review by rgomezcasas — CHANGES_REQUESTED** [review]

File: dotfiles_template/shell/exports.sh [hunk 1]
```diff
@@ -1,3 +1,6 @@
+export DOTLY_AUTO_UPDATE_DAYS=7
+export DOTLY_AUTO_UPDATE_MODE="reminder" # reminder*, auto, prompt
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply dotfiles_template/shell/exports.sh:2]: Done.

Now have all these modes:
- `silent` which update automatically and silently.
- `reminder` which just print a reminder.
- `minor` which update only minor or patch version (need some more work). It is production ready but needs a file "$DOTLY_PATH/VERSION" with current semver version of installed dotly.
- `info` prints a message in terminal and inbox in prompt
- `prompt` just see a inbox in your prompt when there is any update.

File: scripts/core/output.sh [hunk 2]
```diff
@@ -5,20 +5,48 @@ green='\033[0;32m'
...
+    values="Y/n"
+    default_check="Yy"
+  else
+    values="y/N"
+    default_check="Nn"
+  fi
+
+  output::question "$question? [$values]" "PROMPT_REPLY"
+  [[ "${PROMPT_REPLY:-$default}" =~ ^[$default_check] ]]
+}
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply scripts/core/output.sh:49]: Solved

File: scripts/dotfiles/create [hunk 3]
```diff
@@ -3,18 +3,25 @@
 set -euo pipefail
 
 source "$DOTLY_PATH/scripts/core/_main.sh"
+source "$DOTLY_PATH/scripts/core/templating.sh"
 
 ##? Create the dotfiles structure
 ##?
 ##? Usage:
-##?    create [dotfiles_path]
+##?    test
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply scripts/dotfiles/create]: 🙏

File: scripts/self/utils/init.sh [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+
+# Needed variables
+DOTLY_INIT_SCRIPTS_PATH=${DOTLY_INIT_SCRIPTS_PATH:-$DOTLY_PATH/shell/init-scripts}
+DOTFILES_INIT_SCRIPTS_PATH=${DOTFILES_INIT_SCRIPTS_PATH:-$DOTFILES_PATH/shell/init-scripts}
+ENABLED_INIT_SCRIPTS_PATH=${ENABLED_INIT_SCRIPTS_PATH:-$DOTFILES_PATH/shell/init-scripts.enabled}
+
+# Check if init script exists in dotly or dotfiles
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply scripts/self/utils/init.sh]: removed

File: scripts/symlinks/apply [hunk 5]
```diff
@@ -18,8 +18,21 @@ symlinks::apply() {
...
 ##?    apply
 ##?
+##? Options:
+##?   -h --help     Show this help
+##?   -v --version  Show the program version
+##?
 docs::parse "$@"
 
+SCRIPT_NAME="dot symlinks apply"
+SCRIPT_VERSION="1.0.0"
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply scripts/symlinks/apply:28]: If I move this to docbloc the script will parse itself to print the current version of the script. Do you want that?

File: shell/bash/themes/codely.sh [hunk 6]
```diff
@@ -8,10 +8,15 @@ codely_theme() {
...
+  UPDATE_MESSAGE=""
 
   if [ $LAST_CODE -ne 0 ]; then
     STATUS_COLOR=$RED_COLOR
   fi
 
-  export PS1="\[\e[${STATUS_COLOR}m\]{\[\e[m\]${MIDDLE_CHARACTER}\[\e[${STATUS_COLOR}m\]}\[\e[m\] \[\e[33m\]${current_dir}\[\e[m\] "
+  if [ -f "$DOTFILES_PATH/.dotly_update_available" ]; then
+    UPDATE_MESSAGE="📬  | "
+  fi
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply shell/bash/themes/codely.sh]: Done.

File: shell/zsh/themes/prompt_codely_setup [hunk 7]
```diff
@@ -19,7 +19,13 @@ prompt_codely_precmd() {
 }
 
 prompt_codely_setup() {
-  local prompt_codely_status='%(?:%F{green}{%F{white}▸%F{green}}:%F{red}{%F{white}▸%F{red}})'
+  local UPDATE_MESSAGE prompt_codely_status
+
+  if [ -f "$DOTFILES_PATH/.dotly_update_available" ]; then
+    UPDATE_MESSAGE="📬 | "
+  fi
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply shell/zsh/themes/prompt_codely_setup:24]: Done

**Comment by gtrabanco** [comment]
# Autoupdater

Now have all these modes:
- `silent` which update automatically and silently.
- `reminder` which just print a reminder.
- `minor` which update only minor or patch version (need some more work). It is production ready but needs a file "$DOTLY_PATH/VERSION" with current semver version of installed dotly.
- `info` prints a message in terminal and inbox in prompt
- `prompt` just see a inbox in your prompt when there is any update.

File: scripts/core/async.sh [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+setInterval() {
+    
+    local command="$1"
+    local after="$2"
+    
+    read -d " " comm <<<"${command}"
+    
+
+
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply scripts/core/async.sh]: * Deleted all not needed lines and not needed functions.
* Solved a bug about print always 1 in screen when executing async.

File: dotfiles_template/shell/paths.sh [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  "/usr/local/opt/ruby/bin"
+  "/usr/local/opt/python/libexec/bin"
+  "/opt/homebrew/bin"
+  "/usr/local/bin"
+  "/usr/local/sbin"
+  "/bin"
+  "/usr/bin"
+  "/usr/sbin"
+  "/sbin"
+)
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply dotfiles_template/shell/paths.sh:18]: Done

File: dotfiles_template/shell/exports.sh [hunk 10]
```diff
@@ -1,3 +1,6 @@
+export DOTLY_AUTO_UPDATE_DAYS=7
```
[the remark this answers is not in the document]
Comment (gtrabanco) [reply dotfiles_template/shell/exports.sh:1]: Done

**Comment by gtrabanco** [comment]
Please check function `files::check_if_path_is_older` for me works great as it was. See in the files the review I did @rgomezcasas

**Comment by gtrabanco** [comment]
# Important
Please check Bash if it works for you and load correctly the prompt. For me dot not show the right prompt but works pretty well all the other stuff.

**Pull request closed without merging**